bin/install-pr

```diff
@@ -9,63 +9,193 @@
+branch=$(curl -sL "$linkerd2_pulls_url/$pr" | jq -r '.head.ref')
+artifacts=$(curl -sL "$linkerd2_kind_integration_url/runs?branch=$branch" | jq -r '.workflow_runs[0].artifacts_url')
+archive=$(curl -sL -H "$auth" "$artifacts" | jq -r '.artifacts[0].archive_download_url')
 
-branch=$(curl -s -H "$auth" "https://api.github.com/repos/linkerd/linkerd2/pulls/$pr" | jq -r '.head.ref')
-
-artifacts=$(curl -s -H "$auth" "https://api.github.com/repos/linkerd/linkerd2/actions/workflows/kind_integration.yml/runs?branch=$branch" | jq -r '.workflow_runs[0].artifacts_url')
-
-archive_url=$(curl -s -H "$auth" "$artifacts" | jq -r '.artifacts[0].archive_download_url')
-
-dir=$(mktemp -d -t "linkerd-pr-$pr.XXXXXXXXXX")
+bindir=$( cd "${BASH_SOURCE[0]%/*}" && pwd )
+rootdir=$( cd "$bindir"/.. && pwd )
+archive_tmp=$(mktemp -d -t "linkerd-pr-$pr.XXXXXXXXXX")
 
-cd "$dir" || exit
-
-echo "### Downloading images ###"
-
-curl -L -o archive.zip -H "$auth" "$archive_url"
+# Change directory temporarily and download artifact.
+cd "$archive_tmp"
 
+curl -L -o archive.zip -H "$auth" "$archive"
 unzip -o archive.zip
 
-echo "### Pushing images ###"
-
-for archive in cli-bin cni-plugin controller debug grafana proxy web
-do 
-  image=$(docker load -i "$archive.tar" | sed "s/Loaded image: //")
-  docker push "$image"
-  tag=$(echo "$image" | cut -f 2 -d ":")
-done
+# Install the Linkerd CLI.
+image=$(docker load -i cli-bin.tar | sed 's/Loaded image: //')
+container_id=$(docker create "$image")
+
+case $(uname) in
+  Darwin)
+    platform="darwin"
+    ;;
+  Linux)
+    platform="linux"
+    ;;
+  *)
+    platform="windows"
+    ;;
+esac
+
+docker cp "$container_id":/out/linkerd-"$platform" "$rootdir"/target/cli/"$platform"/linkerd
+installed_cli=1
+
+# Load images into local docker images
+if [ $is_kind ]
+then
+  # Create new KinD cluster
+  cluster_name="pr-$pr"
+  context_flag=(--context kind-"$cluster_name")
+
+  if "$bindir"/kind get clusters | grep "$cluster_name"
+  then
+    echo "Error: KinD cluster '$cluster_name' exists"
+    exit 1
+  fi
+  
+  "$bindir"/kind create cluster --name "$cluster_name"
+
+  # Load images into cluster
+  for image in cni-plugin controller debug grafana proxy web
+  do
+    echo "Loading $image image into cluster.."
+    "$bindir"/kind load image-archive --name "$cluster_name" $image.tar || tee load_fail &
+  done
+
+  # Wait for `kind load` background processes to copmlete. Exit early if any
+  # job failed.
+  wait < <(jobs -p)
+  if [ -f load_fail ]
+  then
+    echo "Loading docker images into KinD cluster failed."
+    exit 1
+  fi
+  kind_setup=1
+else
+  for image in cni-plugin controller debug grafana proxy web
+  do
+    docker load -i "$image.tar"
+  done
+fi
 
 cd -
 
-rm -rf "$dir"
-
-linkerd=$(bin/docker-pull-binaries "$tag" | head -n 1)
-
-echo "### Pre checks ###"
-
-$linkerd check --pre
-
-echo "### Installing $tag ###"
-
-$linkerd install | kubectl apply -f -
-
-$linkerd check
+if [ $is_kind ]
```

Comment:
if  the context flag is just empty for the non-kind case, I think we can remove this conditional

Reply:
Unfortunately this is like this for a reason. If `context_flag` is empty, it's passed as an empty string option `""`. I have these separated so that that is not an issue.

If you don't quote the `${context_flag[@]}` it "works", but then `shellcheck` fails with an error.

Reply:
I took a stab at trying to clean this up, but to no avail.  Bash is insane.  We may want to just add a shellcheck exclusion for this since we know we don't have to worry about re-globbing in this case.

Reply:
Disabled these `shellcheck` checks 👍